core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -460,7 +460,8 @@ class Partition(val topic: String,
     }.map(_.logEndOffset)
     val newHighWatermark = allLogEndOffsets.min(new LogOffsetMetadata.OffsetOrdering)
     val oldHighWatermark = leaderReplica.highWatermark
-    if (oldHighWatermark.messageOffset < newHighWatermark.messageOffset || oldHighWatermark.onOlderSegment(newHighWatermark)) {
+    if (oldHighWatermark.messageOffset < newHighWatermark.messageOffset ||
+      (oldHighWatermark.messageOffset == newHighWatermark.messageOffset && oldHighWatermark.onOlderSegment(newHighWatermark))) {
```

Comment:
Can you explain this addition? Is there a scenario where the new high watermark might be smaller than the old high watermark, but be on a newer segment?

Reply:
During my investigation of the negative hw, it appears that we have `LogOffsetMetadata.UnknownOffsetMetadata=new LogOffsetMetadata(-1, 0, 0)`, whereas Replica.highWatermarkMetadata is initialized to be `LogOffsetMetadata(0, -1, -1)`. So strictly speaking we can have two instances LogOffsetMetadata where the messageOffset of one instance is larger and the segmentBaseOffset of the other instance is larger.

After further investigation, I realized that this is not the root cause of the negative hw. But it still seems safer to add this additional check to make sure that we never decrease hw.

Reply:
Note that `allLogEndOffsets.min(new LogOffsetMetadata.OffsetOrdering` in `Partition.maybeIncrementLeaderHW` may assign `LogOffsetMetadata.UnknownOffsetMetadata=new LogOffsetMetadata(-1, 0, 0)` as `newHighWatermark` in the case that the new leader has not received FetchRequest from some follower.

Reply:
Thanks for the explanation. I agree it is safer with the additional check.

Reply:
Is it worth adding a comment here? It's a bit confusing otherwise.

Reply:
I added this comment:
`Ensure that watermark offset does not increase even if segmentBaseOffset of the newer LogOffsetMetadata is larger`